---
title: FAQs
permalink: faqs/
nav_order: 7
---
# Frequently Asked Questions
{: .no_toc }

1. TOC
{:toc}

## Can I use my consumer genetic testing data (23andMe / Ancestry.org / etc...) with PharmCAT?

[PharmCAT requires input genomic data to be in VCF format.](specifications/VCF-Requirements) If you can transform your data into valid VCF that meets the requirements outlined in PharmCAT documentation then you can run it. However, if you're not familiar with genomic data tools then this may be an extremely difficult task.

PharmCAT does not test against datasets generated by consumer testing companies so we can make no claim about how well they work. Many consumer testing datasets have limited overlap with most of the gene definitions used by PharmCAT so this could result in results with very few callable alleles and, thus, not very useful reports. PharmCAT works best with data derived from whole genome sequencing (WGS) datasets that have good coverage.

Regarding 23andMe in particular, some things to consider:

1. **23andMe does genotyping and not sequencing**. You will not have full coverage of the positions used by PharmCAT. This means you will have to make assumptions about those missing positions.
2. As of May 2021, **[23andMe uses the GRCh37 assembly](https://customercare.23andme.com/hc/en-us/articles/212883767-Which-reference-genome-does-23andMe-use-)**. This means you will have to re-align your data to the GRCh38 assembly that PharmCAT uses.


## Can PharmCAT treat missing positions as wild type?

PharmCAT will not be supporting this.

We want people to be 100% clear on how PharmCAT works and what happens with the data you provide to it.  It does not accept arbitrary VCF for many reasons (see [VCF Requirements](docs/specifications/VCF-Requirements.md) for the full list of requirements), but the main one is that we will not make any assumptions on the input you provide.  We have already encountered users making assumptions on how PharmCAT works or should work which has led to confusion down the line.

For one thing, we do not know what "wild type" is because it can vary based on your reference sequence. Did you convert it from GRCh37 to GRCh38? If so, the "wild types" from the two could have changed and your VCF would not provide any indications that this is the case.  Secondly, a missing entry can mean that the reference base was detected OR it can mean the base was not assayed or has no call.  We cannot distinguish between uncalled positions and reference in a VCF file. So we ask that you declare each required position for PharmCAT to be clear about the input.
  
You have to decide on how accurate you want the data you provide to PharmCAT should be, especially if you're making any clinical decisions based on PharmCAT's results.  If you wish to make assumptions of your data, you are welcome to do so.  Instructions on how to do this can be found [here](docs/specifications/Preparing-VCF-Files.md).


## How can I get updates about PharmCAT?

The PharmCAT project is managed in GitHub which has many features for people who want to stay aware of changes happening with PharmCAT.

First, [sign up for a free GitHub account](https://github.com/).

Second, go to the [PharmCAT repository](https://github.com/PharmGKB/PharmCAT), click the watch button, and [configure notifications in a way that works for you](https://docs.github.com/en/enterprise-cloud@latest/account-and-profile/managing-subscriptions-and-notifications-on-github/setting-up-notifications/configuring-notifications).


## When does PharmCAT release new versions?

PharmCAT releases new versions when substantial updates are ready to be released and not on a time-based schedule. For more information see the [Versioning documentation](methods/versioning).