---
title: Home
permalink: /
nav_order: 1
---

# PharmCAT:<br />Pharmacogenomics Clinical Annotation Tool

<span style="float: right;">
![PharmCAT logo](images/pharmcat_logo.png)
</span>

[Download {{ site.github.latest_release.tag_name }}](https://github.com/PharmGKB/PharmCAT/releases/latest){: .btn .btn-blue .umami--click--download-button } [View on GitHub](https://github.com/PharmGKB/PharmCAT){: .btn }

An active area of genomic medicine implementation at many health care organizations and academic medical centers includes development of decision support and return of results around pharmacogenomics.  One of the challenges in implementing pharmacogenomics is the representation of the information in clinical dosing guidelines, including star-allele haplotypes, and extracting these variants and haplotypes from genetic datasets.  In a collaboration between the [Pharmacogenomics Knowledgebase (PharmGKB)](https://www.pharmgkb.org) and the former [PGRN Statistical Analysis Resource (P-STAR)](http://www.pgrn.org/p-star.html), with input from other groups, we are developing a software tool to extract guideline variants from a genetic dataset (represented as a vcf), interpret the variant alleles, and generate a report with genotype-based prescribing recommendations which can be used to inform treatment decisions.

The [Clinical Pharmacogenetics Implementation Consortium (CPIC)](https://cpicpgx.org) has established guidelines surrounding gene-drug pairs that can and should lead to treatment modifications based on genetic variants.  These guidelines are used for the initial version of PharmCAT, and other sources of PGx information and guidelines will be included in the future.

References:
- _Commentary:_ TE Klein, MD Ritchie. [PharmCAT: A Pharmacogenomics Clinical Annotation Tool](https://dx.doi.org/10.1002/cpt.928). Clinical Pharmacology & Therapeutics (2018) 104(1):19-22.
- _Methods paper:_ K Sangkuhl & M Whirl-Carrillo, et al. [Pharmacogenomics Clinical Annotation Tool (PharmCAT)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6977333). Clinical Pharmacology & Therapeutics (2020) 107(1):203-210.

PharmCAT is under active development.

## Contact

[Ask a Question](mailto:pharmcat@pharmgkb.org){: .btn .btn-blue .umami--click--contact } [Submit a Bug / Feature Request](https://github.com/PharmGKB/PharmCAT/issues/new){: .btn .umami--click--bug }

## Documentation

[Summary of Genes and Drugs included in PharmCAT](/Genes-Drugs). There are detailed documents on how particular genes are handled by PharmCAT. See the [gene definition exceptions](methods/Gene-Definition-Exceptions) for a rundown of exceptional circumstances when analyzing particular genes.

Read [the Methods](methods) to learn how PharmCAT works, including how it [matches genotype data to allele definitions](methods/NamedAlleleMatcher-101).

## Running PharmCAT

[Learn how to run PharmCAT](using) and all the different components that make up the PharmCAT tool.  Please make sure to also read and understand PharmCAT's [VCF requirements](specifications/VCF-Requirements).

## Examples

We have an example collection of synthetic input and output files generated by PharmCAT.

[View examples of input and output files](examples).
